README

** Windows systems:

GPU parallelization is not  
currently available on Windows systems.


** Unix systems:

For full operational capabilities,
the 'uroot' package requires the following installed on the system:

  1) CUDA capable GPU with compute capability >= 3.0.
  
  2) CUDA software, which includes the 'nvcc' (release >= 7.1)
     NVIDIA Cuda Compiler driver (available at 
     http://www.nvidia.com/cuda).

  3) A general purpose C compiler is needed by nvcc.

By default the package is installed without the GPU capabilities.  To
request them, set environment variable CUDA_IGNORE to any nonempty value
for the R package installer.


NEWS
# uroot 2.1-1 2020-03-26
-------------------------

* corrected a wrong value in the internal dataset `.CH.orig.cv`. Element
  `uroot:::.CH.orig.cv[12, 6]` was `2.510`, now it is `3.510` as in the original
  source (Canova and Hansen 1995, Table 1), reported by Javier Lopez de Lacalle.


# uroot 2.1-0 2019-08-19
*------------------------

* removed all stuff related to CUDA, since the configuration files
  need more than the tweeks by the current maintainer in versions
  2.0.10 - 2.0.11.


# uroot 2.0-11 2019-08-18
*------------------------

* in configuration files, replaced non-POSIX options '-maxdepth' (find) and '-V'
  (sort) with equivalent POSIX approximations.


# uroot 2.0-10 2019-06-17
-------------------------

* new maintainer (Georgi N. Boshnakov).

* changed the configuration files to pass cleanly CRAN checks.  The changes are
  minimal and mainly concern the non-GPU compilation and symbol registration.  I
  didn't make any structural changes except that now the default installation is
  without GPU support, even if it seems present (set environment variable
  CUDA_IGNORE to force using the GPU). The main reason for this is that the
  compilation may fail since the nvcc compiler has rather strict system
  requirements, which are different for different combinations of versions of
  nvcc, flavours of OS, and OS versions. These may conflict with the tools used
  by the R tool chain.  Currently, as in the previous version of 'uroot',
  installation of GPU code on Windows is not attempted.

* deactivated Rd documentation links to the author's website, www.jalobe.com,
  since it seems unreachable. The vignette is unchanged though.


# uroot 2.0-9 2017-01-27
-------------------------

  o Further improvements to make configure portable. 
    Following the suggestion by Kurt Hornik, I went back to the 
    initial idea of defining a rule %cu : %.o (as those defined in    
    /usr/lib/R/etc/Makeconf) and let Makeconf do the rest.
    The problem I found before was that the default linker (e.g. g++) 
    couldn't link objects compiled with nvcc. Now, I have found that 
    a second stage "nvcc -dlink" is needed in order to generate 
    objects to be linked by a linker other than nvcc.
    Required flags are passed to the linker via PKG_LIBS.

    When cuda files are compiled, R CMD reports 
    "Found ‘exit’, possibly from ‘exit’ (C)"
    This may require some fix in the future; currently 
    no issue was observed when using the package.

# uroot 2.0-8 2017-01-24
-------------------------

  o Configure failed on some systems. As reported by CRAN maintainer:
    """
    on Fedora we see
    ** libs
    g++ -std=gnu++98 -shared -L/usr/local/lib64 -o uroot.so
    g++: fatal error: no input files
    and the corresponding .mk with
    uroot:
       $(CXX) -shared $(LDFLAGS) -o uroot.so
    is simply invalid: not only are there no source files, '-shared' is not
    portable and CXXFLAGS and CXXPICFLAGS are missing.
    """
    Current solution: define a NAMESPACE.in where "dynload(uroot)" is 
    included at the configure step depending on whether CUDA is detected or not.
    configure.win remains as before (remove directory 'src', as I cannot check
    the configure step on a Windows system with CUDA).
    As reported by R CMD check, "NAMESPACE.in" is a non-standard file, 
    but it seems the easiest way to avoid a dummy .so file.

# uroot 2.0-7 2017-01-22
-------------------------

  o The configure script has been modified slightly in the hope 
    to be more portable. 
    
  o The flag -lR has been removed when the dummy uroot.so file is 
    generated (when CUDA is not available).

  o configure.win has been simplified. CUDA files are ignored because 
    I couldn't test it on a windows system with a CUDA enabled GPU.

  o Added document uroot-intro.pdf with links to further documentation.

# uroot 2.0-6 2017-01-05
-------------------------

  o A cleanup script has been added.

  o GNU make is no longer a SystemRequirements.
    Include files are used in order to avoid the 'ifeq' GNU make extension,
    this makes Makevars more portable.

  o The configure script creates now appropriate environment variables 
    when it detects that it is running on the ARINA cluster of the UPV/EHU.

# uroot 2.0-5 2016-03-18
-------------------------

  o Based on an older version of package uroot.
    This version provides the CH test for seasonal stability and
    the HEGY test for seasonal unit roots. The functions that implement 
    these tests have been coded from the scratch in order to include the  
    following new features:
    1) the tests are applicable to series of any seasonal periodicity,
    2) p-values based on response surface regressions are available for both tests,
    3) bootstrapped p-values are available for the HEGY test.


R/hegy-test.R

hegy.test <- function(x, deterministic = c(1,0,0), 
  lag.method = c("fixed", "AIC", "BIC", "AICc"), maxlag = 0,
  pvalue = c("RS", "bootstrap", "raw"), 
  #args related to response surface method
  rs.nobsreg = 15, 
  #args related to bootstrap method
  boot.args = list(seed = 123, 
    #lag.method = c("fixed", "AIC", "BIC", "AICc"), maxlag = 0,
    lag.method = lag.method[1], maxlag = maxlag, 
    byseason = FALSE, nb = 1000, BTdim = c(100, 10), debug.tid = -1))
{
  F.test.stat <- function(id)
  {
    if (length(id) == S)
    {
      if (is.null(xreg))
      {
        #model with no HEGY regressors and deterministic=c(0,0,0)
        rss1 <- sum(dx^2)
        dfs1 <- length(dx)
        return(((rss1 - rss2) / (dfs1 - dfs2)) / (rss2 / dfs2))
      } else
        fit1 <- lm(dx ~ 0 + xreg)
    } else
      fit1 <- lm(dx ~ 0 + cbind(xreg, ypi[,-id]))
    rss1 <- deviance(fit1)
    dfs1 <- df.residual(fit1)
    ((rss1 - rss2) / (dfs1 - dfs2)) / (rss2 / dfs2)
  }

  xreg <- NULL
  stopifnot(is.ts(x))
  n <- length(x)
  S <- frequency(x)
  if (S < 2)
    stop("the time series is not seasonal")
  isSeven <- (S %% 2) == 0
  isSevenp2 <- 2 + isSeven
  dx <- diff(x, S)

  #xreg <- if (is.matrix(xreg)) xreg[-seq_len(S),] else xreg[-seq_len(S)]
  if (!is.null(xreg) && NROW(xreg) != length(dx))
      stop("the number of rows in argument ", sQuote("xreg"), 
        " does not match the length of ", sQuote("diff(x, frequency(x))"))

  data.name <- deparse(substitute(x))
  lag.method <- match.arg(lag.method)
  isNullxreg <- is.null(xreg)
  pvalue <- match.arg(pvalue)

  if (lag.method == "AICc" && pvalue == "RS")
  {
    ##NOTE 
    #response surfaces are not provided by the reference paper DE14
    #they are available for the Hannan-Quinn criterion, see add this option here
    lag.method <- "AIC"
    warning("argument ", sQuote("lag.method"), " was changed to ", sQuote("AIC"))
  }

  if (pvalue == "bootstrap")
  {
    #if (!isNullxreg)
    #  warning("argument ", sQuote("xreg"), " was set to NULL")

    #if (!is.null(boot.args$lag.method))
    #boot.args$lag.method <- boot.args$lag.method[1]

    # default arguments
    #bargs <- list(seed = 123, lag.method = "fixed",
    #  maxlag = 0, byseason = FALSE, nb = 1000, BTdim = c(100, 10), debug.tid = -1)
    bargs <- list(seed = 123, lag.method = lag.method[1],
      maxlag = maxlag, byseason = FALSE, nb = 1000, BTdim = c(100, 10), debug.tid = -1)

    nms1 <- names(boot.args)
    notinldef <- !(nms1 %in% names(bargs))
    wno <- which(notinldef)

    if (any(notinldef))
      warning("the following elements in ", sQuote("boot.args"), " were omitted: ", 
      paste(sQuote(nms1[wno]), collapse = ", "), ".", call. = FALSE)

    if (length(boot.args) > 0)
    {
      if (any(notinldef)) {
        bargs[nms1[which(!notinldef)]] <- boot.args[-wno] 
      } else
        bargs[nms1] <- boot.args
    }
  }

  # storage matrix

  stats <- matrix(nrow = floor(S/2) + 3, ncol = 2)
  id <- seq.int(isSevenp2, S, 2)
  rownames(stats) <- c("t_1", if (isSeven) "t_2" else NULL,
  paste("F", apply(cbind(id, id + 1), 1, paste, collapse=":"), sep = "_"), 
  paste0("F_2:", S), paste0("F_1:", S))
  colnames(stats) <- c("statistic", "p-value")

  # deterministic terms

  if (all(deterministic == c(0,0,1)))
    deterministic <- c(1,0,1)

  if (deterministic[1] == 1) {
    strdet <- "constant"
    xreg <- cbind(xreg, c = rep(1, n - S))
  } else strdet <- ""

  ##NOTE
  #in principle it is not a good idea to define the intercept in "xreg" and 
  #use lm(x ~ 0 + xreg) because stats::summary.lm uses attr(z$terms, "intercept")
  #to compute the R-squared and the F statistic includes the intercept (not the usual test),
  #but here the R-squared and the F-statistic are not used

  if (deterministic[2] == 1) {
    xreg <- cbind(xreg, trend = seq_along(dx))
    strdet <- paste(strdet, "+ trend")
  }

  if (deterministic[3] == 1)
  {
    SD <- do.call("rbind", replicate(ceiling(length(dx)/S), diag(S), simplify = FALSE))
    SD <- ts(SD, frequency = S, start = start(dx))
    # ignore warning "'end' value not changed"
    SD <- suppressWarnings(window(SD, start = start(dx), end = end(dx)))
    colnames(SD) <- paste0("SD", seq_len(S))
    if (deterministic[1] == 1)
      SD <- SD[,-1]
    xreg <- cbind(xreg, SD)
    strdet <- paste0(strdet, ifelse(deterministic[1] == 1, " + ", ""), "seasonal dummies")
  }

  if (strdet == "")
    strdet <- "none"

  # lags of the dependent variable

  if (maxlag > 0)
  {
    dxlags <- dx
    for (i in seq_len(maxlag))
      dxlags <- cbind(dxlags, lag(dx, -i))
    dxlags <- window(dxlags, end = end(dx))[,-1]
    if (is.null(dim(dxlags)))
      dxlags <- as.matrix(dxlags)
    #NOTE this way would be a bit faster 
    #sapply(seq_len(maxlag), function(x) c(rep(NA,length.out=x), c(1:10)[1:(10-x)]))
    colnames(dxlags) <- paste("Lag", seq_len(maxlag), sep ="")

    if (lag.method == "fixed")
    {
      #NOTE 'data.frame' keeps column names, 'cbind' doesn't;
      #'data.matrix' is required in order to be accepted by 'lm'
      xreg <- data.matrix(data.frame(xreg, dxlags))
    } # else, do not add dxlags here since they will be selected below

  } #else # maxlags == 0 #nlags == 0
    #str.lag.order <- lag.order <- 0

  # HEGY regressors

  ypi <- hegy.regressors(x)

  # lag order selection

  if (maxlag > 0)
  {
    if (lag.method != "fixed")
    {
      # information criteria are obtained using the same number of observations
      # regardless of the number of lags included in a given model
      tmp <- vector("list", maxlag + 1)
      # model with no lags
      id <- seq_len(maxlag)
      tmp[[1]] <- lm(dx[-id] ~ 0 + ypi[-id,] + xreg[-id,])
      dxlags2 <- dxlags[-id,]

      #http://stackoverflow.com/questions/23154074/writing-a-loop-in-r-that-updates-a-model?rq=1
      #the loop does not resolve the "i" and new variables are not distinguished
      #tmp[[i+1]] <- update(tmp[[i]], . ~ . + dxlags[,i,drop=FALSE])
      for (i in seq_len(maxlag))
        tmp[[i+1]] <- update(tmp[[i]], as.formula(paste0(". ~ . + dxlags2[,",i,"]")))

      icvals <- unlist(switch(lag.method, 
        "AIC" = lapply(tmp, AIC), "BIC" = lapply(tmp, BIC),
        "AICc" = lapply(tmp, function(x) { k <- x$rank+1; -2*logLik(x) + 2*k + 
          (2*k*(k+1))/(length(residuals(x))-k-1) })))
      id <- which.min(icvals)
      # update lag order (tmp[[1]] contains 0 lags)
      maxlag <- id - 1
      if (maxlag > 0)
      {
        xreg <- data.matrix(data.frame(xreg, dxlags[,seq_len(maxlag),drop=FALSE]))
      }
    } # else, lag.method=="fixed", dxlags was added to reg above
  }

  # fit the model with the chosen lags (if any) 
  # using the entire sample (i.e., including the first 'maxlag' observations)
  fit2 <- lm(dx ~ 0 + ypi + xreg)
  
  # test statistics

  rss2 <- deviance(fit2)
  dfs2 <- df.residual(fit2)

  j <- isSevenp2
  for (i in seq.int(isSevenp2, S, 2))
  {
    id <- c(i, i+1)
    stats[j,1] <- F.test.stat(id)
    j <- j + 1
  }

  stats[j,1] <- F.test.stat(seq.int(2, ncol(ypi)))

  # F-test statistic for all HEGY regressors
  stats[j+1,1] <- F.test.stat(seq_len(S))

  # t-test statistics
  id <- seq_len(1 + isSeven) #+ ncxreg
  stats[seq_along(id),1] <- coef(fit2)[id] / sqrt(diag(vcov(fit2))[id])

  # p-values

  # used with pvalue equal to "RS" or "raw"
  ctd <- paste(deterministic, collapse = "")

  if (pvalue == "RS")
  {
    stats[1,2] <- hegy.rs.pvalue(stats[1,1], "zero", ctd, lag.method, 
      maxlag, S, dfs2, rs.nobsreg)
    if (isSeven)
      stats[2,2] <- hegy.rs.pvalue(stats[2,1], "pi", ctd, lag.method, 
        maxlag, S, dfs2, rs.nobsreg)

    for (i in seq.int(isSevenp2, (S+4)/2-1))
      stats[i,2] <- hegy.rs.pvalue(stats[i,1], "pair", ctd, lag.method, 
        maxlag, S, dfs2, rs.nobsreg)

    j <- nrow(stats) - 1
    stats[j,2] <- hegy.rs.pvalue(stats[j,1], "seasall", ctd, lag.method, 
      maxlag, S, dfs2, rs.nobsreg)
    stats[j+1,2] <- hegy.rs.pvalue(stats[j+1,1], "all", ctd, lag.method, 
      maxlag, S, dfs2, rs.nobsreg)
  } else 
  if (pvalue == "bootstrap")
  {
    e <- residuals(fit2)
    dgp.nlags <- maxlag
    if (dgp.nlags > 0)
    {
      arcoefs <- coef(fit2)
      arcoefs <- arcoefs[grepl("Lag\\d{1,3}$", names(arcoefs))]
      if (length(arcoefs) != dgp.nlags) # debug
        stop("unexpected number of lags")
    } else
      arcoefs <- 0

    BTdim <- bargs$BTdim
    nb <- bargs$nb
    if (is.null(BTdim)) {
      BTdim <- rep(ceiling(sqrt(nb)), 2)
    } else
    if (prod(BTdim) < nb) {
      stop("the number of threads is lower than the number of bootstrap replicates")
    }

    #NOTE n-S is passed as the length of the data, 
    #the length of diff(x,S) is required by hegy_boot_pval

    if (bargs$byseason)
    {
      e <- ts(e, frequency = S)
      csns <- cumsum(table(cycle(e)))
      #names(csns) <- cycle(dx)[seq.int(dgp.nlags+1, length.out=S)]

      e <- unlist(split(e, cycle(e)))

    } else {
      csns <- c(0, length(e))
    }

    boot.lag.method <- match.arg(bargs$lag.method, c("fixed", "AIC", "BIC"))
    chosen.lags <- if (boot.lag.method == "fixed")
      integer(1) else integer(bargs$nb)

    ## removed with cuda stuff
    ## if (@HAS_CUDA@)
    ## {
    ##   tmp <- .C("hegy_boot_pval", dim = as.integer(c(nb, BTdim)), 
    ##     seed = as.integer(bargs$seed), idc = as.integer(deterministic), 
    ##     ip = as.integer(c(S, n-S, dgp.nlags, bargs$maxlag, bargs$debug.tid)),
    ##     csns = as.integer(csns),
    ##     eps = as.double(e), arcoefs = as.double(arcoefs), 
    ##     stats0 = as.double(stats[,1]), 
    ##     ICtype = as.integer(switch(boot.lag.method, 
    ##       "fixed" = 0, "AIC" = 1, "BIC" = 2, "AICc" = 3)), 
    ##     bpvals = double(nrow(stats)), chosen_lags = chosen.lags, PACKAGE = "uroot")
    ##   stats[,2] <- tmp$bpvals
    ##   chosen.lags <- tmp$chosen_lags
    ## } else {
    ##   if (grepl("windows", .Platform$OS.type)) {
    ##     stop("GPU parallelization could not be tested on a Windows system,\n", 
    ##     "consider using the function ",  sQuote("hegy.boot.pval"))
    ##   } else # unix
    ##     stop("CUDA capable GPU was not available when installing the package on this system,\n", 
    ##     "consider using the function ",  sQuote("hegy.boot.pval"))
    ## }
    stop("'pvalue = bootstrap' was only available with CUDA, but\n",
         "CUDA capablities in package 'uroot' were removed from v. 2.1-0,\n", 
         "consider using the function ",  sQuote("hegy.boot.pval"))

  } else
  if (pvalue == "raw")
  {
    #NOTE documentation deterministic == c(0,0,1) not available 
    #in original tables (use c(1,0,1) instead)
    stats[1,2] <- uroot.raw.pvalue(stats[1,1], "HEGY", NULL, dfs2, ctd, S, "zero")
    if (isSeven)
      stats[2,2] <- uroot.raw.pvalue(stats[2,1], "HEGY", NULL, dfs2, ctd, S, "pi")
    for (i in seq.int(isSevenp2, (S+4)/2-1))
      stats[i,2] <- uroot.raw.pvalue(stats[i,1], "HEGY", NULL, dfs2, ctd, S, "pair")
  }

  # output

  res <- list(statistics = stats[,1], pvalues = stats[,2],
    method = "HEGY test for unit roots", data.name = data.name, 
    fitted.model = fit2, 
    lag.method = lag.method, lag.order = maxlag,
    strdet = strdet, type.pvalue = pvalue,
    bootstrap = if (pvalue == "bootstrap") bargs else NULL,
    boot.chosen.lags = if (pvalue == "bootstrap") chosen.lags else NULL,
    pvlabels = symnum(stats[,"p-value"], corr = FALSE, na = FALSE,
      cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
      symbols   =  c("***","**","*","."," ")))
  class(res) <- "HEGYtest"

  res
}


src/dev-fnc.h
#ifndef _DEV_FUNCTIONS_
#define _DEV_FUNCTIONS_

#include "R.h"

#include <curand_kernel.h>
//#include <stdio.h>

__device__ void boot_dgp_init(curandState *RNG_state, const int S, 
  const int dgp_nlags, const double *arcoefs, const int *csns, 
  const double *eps, double *dylags, double *y0);

__device__ void crossprod_regmatrix(curandState *RNG_state, const double *eps, 
  double *y0, const int nx, const int *csns, const int S, 
  const int idc0, const int idc1, const int idc2, const int nlags, 
  const int dgp_nlags, const double *arcoefs, 
  double *dylags, double *res);

__device__ void sweep(const int nv, const int nc, const int idcol, double *v);

__device__ void hegy_statistics(double *v, const int nv, const int nc, const int nr, 
  const int S, const int dgp_nlags, const double *arcoefs, double *res);

#endif


R/hegy-rs-pvalue.R

# the code for function 'hegy.rs.pvalue' below has been 
# ported from gretl code provided by Ignacio Díaz-Emparanza (2014)
# http://www.ehu.eus/ignacio.diaz-emparanza/packages/GHegy.gfn
# "Numerical Distribution Functions for Seasonal Unit Root Tests", 
# Computational Statistics and Data Analysis, Volume 76, pages 237-247 
# The Annals of Computational and Financial Econometrics - 2nd Issue
# DOI: http://dx.doi.org/10.1016/j.csda.2013.03.006

hegy.rs.pvalue <- function(x, type = c("zero", "pi", "pair", "seasall", "all"), 
  deterministic = c(1,0,0), lag.method = c("fixed", "AIC", "BIC"), #"HQC"
  lag.order, S, n, nobsreg)
{
  type <- match.arg(type)
  if (is.numeric(deterministic))
    deterministic <- paste(deterministic, collapse = "")
  lag.method <- match.arg(lag.method)

  if (deterministic == "000")
    stop("the combination of deterministic components ", sQuote(deterministic), 
      " is not available for ", sQuote("pvalue=\"RS\""))

  ##NOTE 
  #casedet = "000" is not considered

  CMlabel <- switch(type, "zero" = "Ct1", "pi" = "Ct2", "pair" = "CF", "seasall" = "CFs", "all" = "CFt")  
  tmp <- switch(deterministic, "100" = "c", "110" = "ct", "101" = "cD", "111" = "cDt")
  CMlabel <- paste(CMlabel, tmp, sep = "_")
  tmp <- switch(lag.method, "fixed" = "fijo", "AIC" = "AIC", "BIC" = "BIC", "HQC" = "HQC")
  CMlabel <- paste(CMlabel, tmp, sep = "_")

  C <- .HEGY.CM.tabs[[CMlabel]]

  nc <- ncol(C)
  C1 <- C[,seq_len(nc-1)]
  sdC1 <- C[,nc]

  r1 <- c(0.0001, 0.0002, 0.0005, seq_len(10)/1000)
  rq <- c(r1, seq.int(15, 985, 5)/1000, 1 - rev(r1))

  xeplc <- c(1, 1/n, 1/n^2, 1/n^3, lag.order/n, lag.order/n^2, lag.order/n^3,
    lag.order^2/n, lag.order^2/n^2, lag.order^2/n^3,
    lag.order^3/n, lag.order^3/n^2, lag.order^3/n^3, S/n, S/n^2, S/n^3)

  Q1 <- C1 %*% xeplc

  nrq <- length(rq)
  centro <- floor(nobsreg/2) + 1

  isFtest <- !(type %in% c("zero", "pi"))

  if (isFtest) {
    stopifnot(NCOL(Q1) == 1)
    Q1 <- rev(Q1)
  }

  bin <- rep(0, nrq)
  bin[Q1 >= 0] <- 1
  sbin <- sum(bin)

  if (isFtest)
    sdC1 <- rev(sdC1)

  Q1s <- sort(Q1)

  if (x < Q1s[1]) {
    if (isFtest) return(1) else return(0)
  } else if (x > tail(Q1s, 1)) {
    if (isFtest) return(0) else return(1)
  } else 
  {
    masque <- which(x > Q1)
    masque <- if (length(masque) > 0) max(masque) else 1
    if (masque < nrq) {
      mascer <- if ((x - Q1[masque]) < (Q1[masque+1] - x)) masque else masque + 1
    } else mascer <- nrq

    centroup <- nrq - centro + 1
    mascer <- if (mascer <= centro) centro else mascer
    mascer <- if (mascer >= centroup) centroup else mascer

    # local regressions with "nobsreg" observations
    qi <- pri <- si <- rep(NA, nobsreg)
    for (i in seq_len(nobsreg))
    {
      j <- mascer - centro + i
      qi[i] <- Q1[j]
      pri[i] <- rq[j]
      si[i] <- sdC1[j]
    }

    Y <- if (isFtest) qchisq(pri, df = 2) else qnorm(pri)
    X <- cbind(1, qi, qi^2, qi^3)
    co <- lm.fit(X, Y)$coef

    # substitute by FGLS
    Sigma <- matrix(0, nrow = nobsreg, ncol = nobsreg)
    for (i in seq_len(nobsreg))
    {
      for (j in seq_len(nobsreg))
        if (i <= j) {
          Sigma[i,j] <- (si[i]*si[j])*sqrt((pri[i]*(1-pri[j]))/(pri[j]*(1-pri[i])))
        } else 
          Sigma[i,j] <- (si[i]*si[j])*sqrt((pri[j]*(1-pri[i]))/(pri[i]*(1-pri[j])))
    }

    Pinv <- t(solve(chol(Sigma)))
    PY <- Pinv %*% Y
    PX <- Pinv %*% X
    co <- lm.fit(PX, PY)$coef

    valorcomp <- sum(co * c(1, x, x^2, x^3))
    if (isFtest) {
      pval <- pchisq(q = abs(valorcomp), df = 2, lower.tail = FALSE)
    } else
      pval <- pnorm(q = valorcomp, lower.tail = TRUE)
  }

  pval
}
